### ISSUE TITLE
1.15.5 regression : can't rotate Active Directory credentials anymore

### ISSUE THREAD

**Issue by tisonv** [hashicorp/vault#25249]
**Describe the bug**
Can't rotate Active Directory credentials anymore since 1.15.5

**To Reproduce**
1. Go to an AD Role https://....../ui/vault/secrets/..../ldap/roles/static/.../details
2. Click 'Rotate Credentials'

Or

1. With Postman or Powershell, POST on https://..../v1/..../rotate-role/... with `X-Vault-Token header`

**Error** 
```
{
    "errors": [
        "1 error occurred:\n\t* unable to finish rotating credentials; retries will continue in the background but it is also safe to retry manually: LDAP Result Code 201 \"Filter Compile Error\": ldap: invalid characters for escape in filter: encoding/hex: invalid byte: U+002C ','\n\n"
    ]
}
```

**Expected behavior**
Credentials are rotated

**Environment:**
* Vault Server Version (retrieve with `vault status`): `1.15.5`
* Vault CLI Version (retrieve with `vault version`): `Vault v1.15.5 (0d8b67ef63815f20421c11fe9152d435af3403e6), built 2024-01-26T14:53:40Z`
* Server Operating System/Architecture: Windows 2022

Vault server configuration file(s):

```hcl
# Paste your Vault config here.
# Be sure to scrub any sensitive values
ui = true

storage "raft" {
  path = "......\\Data"
  node_id = "XXXXXXX"
}

# HTTPS listener
listener "tcp" {
  address       = "0.0.0.0:8200"
  tls_cert_file = "cert.pem"
  tls_key_file  = "key.pem"
}

api_addr = "http://X.X.X.X:8200"
cluster_addr = "https://X.X.X.X:8201"
```

**Additional context**

- Rotation was working nicely up to 1.15.4
- Vault configuration was not changed during the update
- No available information in log file when launch with `vault server -config=....hcl >> log.txt 2>&1`
- We tried retyping every ',' but no changes.
- We have several ldap url separated by commas. Using only 1 didn't fix the problem.
- User search filter : `({{.UserAttr}}={{.Username}})`
- Group Filter : `(member={{.UserDN}})`
- LDAP login is still working

**Comment by tlskinneriv** [comment]
Having the same issue here, though we are just having an issue rotating the root credential. Our other accounts are using the username field (instead of DN), which appears to be working fine. I suspect what we're seeing is an issue in the filter changes for updated DN type passwords as part of [changes in v0.11.3 of the openldap plugin](https://github.com/hashicorp/vault-plugin-secrets-openldap/compare/v0.11.2...v0.11.3). Haven't had a chance to dig into the code yet to flesh out what exactly it is.

**Comment by raymonstah** [comment]
Hi! Thanks for opening the issue. I'm curious what the `userattr` is, defined in your config?

**Comment by tisonv** [comment]
Thanks for the replies

@raymonstah 
userattr  = `samaccountname`

**Comment by raymonstah** [comment]
Hmm, okay. I'm attempting to debug this. Based on your error message, it seems like the error occurs when the search filter is applied. I'm curious if there are any `,` characters in your static role name.

**Comment by tlskinneriv** [comment]
@raymonstah For us, the userattr = `userPrincipalName`. We haven't have any trouble rotating any other roles except for the root role. IIRC, the root role is only defined by DN (the `binddn` parameter in the config), which would have multiple `,` characters in its identifier if it's the full DN in an Active Directory environment.

**Comment by raymonstah** [comment]
@tlskinneriv Is the error message you received different?
@tisonv thanks for providing that information -- I'll continue investigating!

**Comment by tlskinneriv** [comment]
Starting at `LDAP Result Code 201` it's the same message. I'll get the full message tomorrow.

**Comment by raymonstah** [comment]
@tlskinneriv Were you able to rotate-root before, with `userattr = userPrincipalName`? Can you confirm if you also have a `upndomain` set in your config?

I think I understand why this behavior is happening. With https://github.com/hashicorp/vault-plugin-secrets-openldap/pull/80, we've introduced the ability to rotate-root when `userattr = userPrincipalName`. It expects the `binddn`, `userdn` and `upndomain` to be set, so that it can correctly form the the search filter. Since there were `,` in your binddn, it were being escaped incorrectly and the search would return an error.

**Comment by tlskinneriv** [comment]
It's been a while since I've done it manually, but I presume it would have worked fine. If you're asking if we changed any config, the answer would be not for a while. 

I'll paste the sanitized details of my coding tomorrow when I have access again. I'll also get the last date that a manual rotation was performed and the version of Vault during that time. 

For the latter, is the code making the assumption that the CN of the user's DN matches the user principal name? If so, that's not always a true statement, as the user principal name is not dependent on CN or DN values and can be configured separately, at least it can in Active Directory.

**Comment by tisonv** [comment]
My bad, the role name is just a plain word.
The ldap secret engines contains a `-`.

**Comment by tlskinneriv** [comment]
Here is my ldap config:

```
Key                              Value
---                              -----
anonymous_group_search           false
binddn                           CN=vault-bind,OU=myOU,DC=my,DC=domain,DC=com
case_sensitive_names             false
certificate                      
connection_timeout               30
deny_null_bind                   true
dereference_aliases              
discoverdn                       false
groupattr                        cn
groupdn                          
groupfilter                      (|(memberUid={{.Username}})(member={{.UserDN}})(uniqueMember={{.UserDN}}))
insecure_tls                     false
last_bind_password_rotation      2023-07-05T13:26:07.084343235Z
max_page_size                    2147483647
password_policy                  password_ad
request_timeout                  90
schema                           ad
skip_static_role_import_rotation false
starttls                         false
tls_max_version                  tls12
tls_min_version                  tls12
upndomain                        
url                              ldaps://my.domain.com
use_token_groups                 false
userattr                         userPrincipalName
userdn                           OU=myOU,DC=my,DC=domain,DC=com
userfilter                       ({{.UserAttr}}={{.Username}})
username_as_alias                false
```

I don't have the last history of when a job was run to rotate the root role, which means it was greater than 3 months ago. If the `last_bind_password_rotation` is the date of last root rotation, then it was back in July 2023. We would have been on 1.14.0 at the time. We've had no reports of any failed rotations other than the root account starting with 1.15.5.

**Comment by tisonv** [comment]
If needed, the complete settings
```
Key                             Value
---                             -----
anonymous_group_search          false
binddn                          CN=myUser,OU=myOU,DC=domain,DC=int
case_sensitive_names            false
certificate                     n/a
connection_timeout              0
deny_null_bind                  true
dereference_aliases             n/a
discoverdn                      true
groupattr                       cn
groupdn                         DC=domain,DC=int
groupfilter                     (member={{.UserDN}})
insecure_tls                    false
max_page_size                   -1
request_timeout                 90
starttls                        false
tls_max_version                 tls12
tls_min_version                 tls12
token_bound_cidrs               []
token_explicit_max_ttl          0s
token_max_ttl                   0s
token_no_default_policy         false
token_num_uses                  0
token_period                    0s
token_policies                  []
token_ttl                       0s
token_type                      default
upndomain                       n/a
url                             ldaps://server1,ldaps://server2
use_pre111_group_cn_behavior    false
use_token_groups                false
userattr                        samaccountname
userdn                          DC=domain,DC=int
userfilter                      ({{.UserAttr}}={{.Username}})
username_as_alias               false
```

**Comment by raymonstah** [comment]
Hey folks, I've put up a pull request / branch that attempts to fix the issue you are running into. Do you have the ability to build custom plugins to point to a development branch? If so, would you mind giving it a shot?

No worries if not, I'll try to get this change backported into the next release.

**Comment by raymonstah** [comment]
Just wanted to share that this has been resolved in Vault 1.15.6

**Comment by tisonv** [comment]
Confirmed working !
Thank you very much
